### DISCUSSION TITLE
Question about rollback and changelog entry

### DISCUSSION THREAD

**Discussion by nazeem-soeltan** [flamingock/mongock#48] [Q&A] [answered]
Hi,

I'm wondering, if the driver is transactionable and `transaction` has been set to `true` should a new entry be created in `mongockChangeLog` with `FAILED` state?

I'm having the following example ChangeUnit. What I see is that the Company entity is not persisted (as expected). And the `executeRollback` is not executed. This is as expected, because mongo's transactional kicks in if I'm correct. However, if I check in `mongockChangeLog`, I don't see any entry reflecting that the ChangeUnit has failed. Is this the expected behaviour?

```java
public class AddSampleCompaniesRollback {

    private static final String COMPANY_NAME = "sample-mongock-rollback";

    @Execution
    public void execute(final MongoTemplate mongoTemplate) {
        final var company = Company.builder()
                .name(COMPANY_NAME)
                .build();

        mongoTemplate.save(company);

        throw new RuntimeException("Let's rollback!");
    }

    @RollbackExecution
    public void executeRollback(final MongoTemplate mongoTemplate) {
        log.info("Rolling back creation of company {}", COMPANY_NAME);

        //mongoTemplate.remove(query(where("name").is(COMPANY_NAME)));
    }

}
```

I'm using mongock `5.0.35`, connected with mongodb `4.2.x` and using Spring Data 3.

**application.yml**
```
mongock:
  enabled: true
  transaction-enabled: true
  migration-scan-package:
    - dev.nazeem.mongock.db
```

**Configuration**
```java
public class EnableMongockConfiguration {

    @Bean
    SpringDataMongoV3Driver springDataMongoDriver(final MongoTemplate mongoTemplate) {
        return SpringDataMongoV3Driver.withDefaultLock(mongoTemplate);
    }

    @Bean
    MongoTransactionManager transactionManager(final MongoTemplate mongoTemplate) {
        return new MongoTransactionManager(mongoTemplate.getMongoDatabaseFactory());
    }

}
```

**Comment by dieppa** [comment, marked as answer]
Hello @nazeem-soeltan ,

You are right, when transaction is enabled, every change unit, together with the migration associated, is shipped in one single transaction, meaning that if anything fails the transaction is not committed and you don't see anything in the database and Mongock will execute that change the next time is executed. At that point the `@RollbackExecution` is not executed.

So, why is the `@RollbackExecution` needed and, moreover, why is it mandatory? Well, basically for the `undo` operation, which differs from rollback in the fact that `undo` is applied when the changes are already committed, normally when a deployment is identified as corrupted after being successfully published.

PD: We probably need to add this explanation to our documentation.

Thanks!

**Reply by nazeem-soeltan** [reply]
Thanks!
Sorry for my ignorance, when you talk about undo operation. What do you actually mean. In what context?

By the way, I'm doing a small investigation on the workings of version 5. (We already use 4). So If you'd like, I can make proposals on improving documentation, which you can review.

**Comment by dieppa** [comment]
Hi @nazeem-soeltan ,

yeah, sure, we always appreciate proposals, that makes us better!

To explain the undo operation, let's imagine you have your application, version1 with a change unit(**migration1**). You have your docker image(**image1**) and deploy it in production. Everything is great. 

Then you keep working, add another change unit(**migration2**), build another docker image of your app(**image2**), deploy it into production and good. Smoothly deployed in production. Everything working as expected. 

However, you realise later that change was based on a wrong assumption, so you need to fall back to **image1**. However, if you deploy image1 directly, it will conflict with the database, as it's reflecting the changes in **migration2**.

Here is where **undo** operation comes to help. We need to update our documentation around operations, but you can take a look [here](https://docs.mongock.io/v5/CLI/operations/index.html#undo-up-to-change-pro)
